apps/iiif/manifests/models.py

```diff
@@ -353,3 +350,15 @@ class RelatedLink(models.Model):
     format = models.CharField(max_length=255, choices=Choices.MIMETYPES, blank=True, null=True)
     profile = models.CharField(max_length=255, blank=True, null=True)
     manifest = models.ForeignKey(Manifest, on_delete=models.CASCADE)
+
+@receiver(signals.pre_save, sender=Manifest)
+def set_logo_url(sender, instance, **kwargs):
+    """
+    Sets logo_url for manifests with logo file upload
+    """
+    # For logo image uploads (i.e. NOT remote manifest ingestion), either new or replacing an existing file
+    if instance.logo and instance.logo != None and instance.logo_url != f'{settings.HOSTNAME}/media/{instance.logo}':
```

Comment:
Two things here:
- `instance.logo and instance.logo != None` is redundant, can just be `instance.logo`
- I would say `not instance.logo_url` rather than checking it specifically against this logo URL. We only want to set it here if it's unset, not if it doesn't match a specific URL. The goal here is to make sure `logo_url` is always populated.

Reply:
> * `instance.logo and instance.logo != None` is redundant, can just be `instance.logo`

True – removed the `instance.logo != None` check.


> * I would say `not instance.logo_url` rather than checking it specifically against this logo URL. We always want to set it here if it's unset, regardless of whether it matches a specific URL. The goal here is to make sure `logo_url` is always populated.

My thinking here is that an existing `logo_url` should be updated if an existing `logo` file changes. Is that something we should be checking for, or not something that we anticipate happening?

Reply:
Ah, OK! Good call, I hadn't thought about that. It's likely an edge case, but important to support.

In that case, I would do something like this:

```py
if not instance.pk and instance.logo and not instance.logo_url:
    # new object; set the logo URL from the logo
else:
    # existing object; get the instance before changes and check if the logo has changed
    existing_instance = sender.objects.get(pk=instance.pk)
    if instance.logo and existing_instance.logo != instance.logo:
        # logo changed, set the logo URL from the logo
```

`logo` is a file field, and I'm not sure if file objects can be compared like that with a simple `!=`, so I would check Django docs and/or google how to check equality between filefield values.

Reply:
Thanks! I agree that this is a better way to do this. Testing this out using breakpoints, the `!=` comparison does seem to recognize the same vs. different file objects as intended. Implemented as suggested 👍

Reply:
I think that just compares filenames—which also makes me realize that `f'{settings.HOSTNAME}/media/{instance.logo}'` isn't reliable in case of filename conflicts 😭 (and I think it depends on Django's `MEDIA_ROOT` setting, among other things). Luckily, there is a [`url` property on the file](https://docs.djangoproject.com/en/3.2/ref/models/fields/#django.db.models.fields.files.FieldFile.url) that we can use.

However, I'm also seeing that we can't actually reliably get the URL for the uploaded file until post-save, see the "note" here: https://docs.djangoproject.com/en/4.1/topics/files/#using-files-in-models.

My new thinking is this:

- In the migration, change the `logo_url`-setting code to just:
   ```py
   man.logo_url = man.logo.url
   ```
- Add a new `post_save` signal that just says something like:
   ```py
   if instance.logo and not instance.logo_url:
      instance.logo_url = instance.logo.url
   ```

- In your existing `pre_save` signal:
   ```py
   if instance.pk and instance.logo_url == instance.logo.url:
       instance.logo_url = ""
   ```

That way, the `pre_save` will clear out the `logo_url` if it matches the URL of the file in the `logo` field. Then, the `post_save` will set it to whatever the logo's URL is after save, if it's blank. So regardless of whether you uploaded a new file or it's the same old file, it will get cleared and reset in the `post_save`.

And if it's not blank, AND it doesn't match `instance.logo.url` during pre-save (in other words, you manually set the `logo_url` or it was set from a remote ingest), then it won't be altered.

Let me know if that makes sense!

Reply:
BTW: if you want to revert the migration to try running it and make sure my suggested change works, you can use [this method](https://docs.djangoproject.com/en/4.1/topics/migrations/#reversing-migrations) for something like `python manage.py migrate manifests 0053`.

I also noticed an indentation issue in your migration:

```py
    for man in manifests:
        man.logo_url = f'{settings.HOSTNAME}/media/{man.logo}'
        Manifest.objects.bulk_update(manifests, ['logo_url'])
```

should be

```py
    for man in manifests:
        man.logo_url = f'{settings.HOSTNAME}/media/{man.logo}' #(although this will change now!)
    Manifest.objects.bulk_update(manifests, ['logo_url'])
```

so the bulk update happens outside of the loop.

Reply:
Good catch! Fixed 👍

Reply:
> That way, the `pre_save` will clear out the `logo_url` if it matches the URL of the file in the `logo` field. Then, the `post_save` will set it to whatever the logo's URL is after save, if `logo_url` is blank at that point. So regardless of whether you uploaded a new file or it's the same old file, it will get cleared in `pre_save` and reset in `post_save`.

I think we'd only want to clear the `logo_url` if it _doesn't_ match the URL of the file in the `logo` field (such as when uploading a different logo file to replace an existing one), right? If they match, then there is no need to alter the `logo_url`. Or am I looking at this backwards?


> * Add a new `post_save` signal that just says something like:
>   ```python
>   if instance.logo and not instance.logo_url:
>      instance.logo_url = instance.logo.url
>   ```

Just an observation, but it seems that setting `instance.logo_url = instance.logo.url` does not persist without saving the instance, which in turn triggers another post-save signal call. It doesn't cause an endless loop because `instance.logo_url` evaluates to `true` the second time around, but I wonder if that is a cause for concern, performance-wise.


> I also noticed an indentation issue in your migration:

Good catch! Fixed 👍  I also added a check in the migration to first make sure that the manifest has a logo, as I had a couple in my local DB that didn't have one (I'm guessing this is unlikely in the prod DB, but still possible) and caused the migration to fail.

Reply:
Ok, I've realized my plan here is flawed from the start. Big apologies for all the time and energy sunk into this so far; thank you for patiently thinking through this together, and hopefully it let you get your feet wet with Django!

New thinking: because of the issues we've discovered here with file URLs, we need a way to definitively differentiate between generated logo file URL and manually populated/remote ingested logo URL. The simplest way to do that is, wherever we need to use the logo URL from a manifest, do the following:

`manifest.logo_url or (manifest.logo.url if manifest.logo else "")`

And never populate `logo_url` from just a file upload.

What that means, unfortunately (or fortunately?? 😅 ) is that the migration and the signal handlers are no longer needed. We just simply generate the logo file's URL on the spot if it's required in case there isn't a `logo_url`.

I think this touches:
- `apps/iiif/serializers/manifest.py`: line 109 `"logo": obj.logo_url or (obj.logo.url if obj.logo else ""),`
- maybe the tests in `apps/iiif/manifests/tests/test_views.py` and `apps/iiif/manifests/tests/tests.py` where `logo_url` is referenced (not sure)

Really sorry for wasting your time with migrations and signal handlers when it wasn't necessary!

<details>
<summary>My long-winded thinking to arrive at the new idea</summary>

> I think we'd only want to clear the `logo_url` if it doesn't match the URL of the file in the logo field (such as when uploading a different logo file to replace an existing one), right? If they match, then there is no need to alter the `logo_url`. Or am I looking at this backwards?

It's very counter-intuitive and a bit of a puzzle. If you upload a different logo only using the file field, at `pre_save`, the `instance.logo_url` will still be the old logo's `existing_instance.logo.url`, because the new logo's `logo.url` can't be computed until post-save. (If the old string method worked to compute this then we could have computed it during pre-save, but it's not reliable.)

Therefore, we need a way to indicate to the post-save signal that it should re-generate the logo URL, but we only want to do that if the prior logo URL was definitely generated from the file field and not set by the `logo_url` field manually/through remote ingest. We don't have a way of comparing reliably if the file itself changed, so we have to regenerate it every time if we have a file upload.

If you manually changed `logo_url`, or it came from a remote ingest, then `instance.logo_url` won't match `existing_instance.logo.url`, and in that case we definitely don't want to clear it out because we want to keep the one that was manually entered (as opposed to generating it from the logo file). There is definitely an assumption here that if you did pull the `logo_url` from the remote manifest, but then decided you actually wanted to upload a file logo instead, that you'd have to clear out the `logo_url` manually. In other words, manually entered `logo_url`s or those from remote manifests are assumed to always take precedence over generated ones.

> Just an observation, but it seems that setting `instance.logo_url = instance.logo.`url does not persist without saving the instance, which in turn triggers another post-save signal call. It doesn't cause an endless loop because `instance.logo_url` evaluates to `true` the second time around, but I wonder if that is a cause for concern, performance-wise.

Good point. Actually, I wonder if this will cause an infinite loop in the following way:
- User uploads the same logo file twice
- The logo_url gets cleared in pre-save
- The logo_url gets regenerated in post-save, then `save()` happens
- The logo_url gets cleared in pre-save because it's the same as it was before...
- etc


> I also added a check in the migration to first make sure that the manifest has a logo, as I had a couple in my local DB that didn't have one (I'm guessing this is unlikely in the prod DB, but still possible) and caused the migration to fail.

Good catch!
</details>

Reply:
Thanks again for all your guidance – I'm glad for the Django learning opportunity 😄

What I've ended up with is:

- On remote ingest, populate `logo_url` if manifest contains logo property
- On local ingest, proceed as usual (manifest has `logo` file object if uploaded, or default file if none uploaded, and no `logo_url` is set)
- Serializer will first return `logo_url`, if there is one, then `logo.url` (or "") if not

I did end up using a pre-save signal to clear out an existing `logo_url` if a `logo` file is uploaded, as the subsequently uploaded should replace the initially ingested `logo_url` in that case.

I also kept the migration (to create the `logo_url` model field) but removing the method that sets a `logo_url` for all existing manifests. 

Also, just FYI – my commit history was starting to get messy so I rebased and squashed into one commit.

What do you think?